api/filters/shared_instance_update_filter.go

```diff
@@ -0,0 +1,146 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package filters
+
+import (
+	"encoding/json"
+	"github.com/Peripli/service-manager/pkg/log"
+	"github.com/Peripli/service-manager/pkg/types"
+	"github.com/Peripli/service-manager/pkg/util"
+	"github.com/Peripli/service-manager/pkg/web"
+	"github.com/Peripli/service-manager/storage"
+	"github.com/sirupsen/logrus"
+	"net/http"
+)
+
+const SharedInstanceUpdateFilterName = "SharedInstanceUpdateFilter"
+
+// SharingInstanceFilter validate the un/share request on an existing service instance
+type sharedInstanceUpdateFilter struct {
+	storageRepository storage.Repository
+}
+
+func NewSharedInstanceUpdateFilter(storageRepository storage.Repository) *sharedInstanceUpdateFilter {
+	return &sharedInstanceUpdateFilter{
+		storageRepository: storageRepository,
+	}
+}
+
+func (*sharedInstanceUpdateFilter) Name() string {
+	return SharedInstanceUpdateFilterName
+}
+
+func (sf *sharedInstanceUpdateFilter) Run(req *web.Request, next web.Handler) (*web.Response, error) {
+	var reqServiceInstance types.ServiceInstance
+	err := util.BytesToObjectNoLabels(req.Body, &reqServiceInstance)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceID := req.PathParams["resource_id"]
+	ctx := req.Context()
+
+	// Get instance from database
+	dbPersistedInstanceObject, err := storage.GetObjectByField(ctx, sf.storageRepository, types.ServiceInstanceType, "id", instanceID)
+	if err != nil {
+		return next.Handle(req)
+	}
+	persistedInstance := dbPersistedInstanceObject.(*types.ServiceInstance)
+
+	// Validate plan has not changed - we don't allow changing plan of shared instance, may have existing references
+	if isPlanChanged(persistedInstance, reqServiceInstance) {
+		return nil, util.HandleInstanceSharingError(util.ErrChangingPlanOfSharedInstance, persistedInstance.ID)
+	}
+
+	if reqServiceInstance.Shared == nil {
+		return next.Handle(req)
+	}
+
+	isAsync := req.URL.Query().Get(web.QueryParamAsync)
+	if isAsync == "true" {
+		return nil, util.HandleInstanceSharingError(util.ErrAsyncNotSupportedForSharing, instanceID)
+	}
+
+	logger := log.C(ctx)
+
+	// we cannot use reqServiceInstance in this validation because the struct has default values (like "" for string type properties)
+	err = validateRequestContainsSingleProperty(logger, req.Body, instanceID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get plan object from database, on service_instance patch flow
+	dbPlanObject, err := storage.GetObjectByField(ctx, sf.storageRepository, types.ServicePlanType, "id", persistedInstance.ServicePlanID)
+	if err != nil {
+		return nil, err
+	}
+	plan := dbPlanObject.(*types.ServicePlan)
+
+	if !plan.IsShareablePlan() {
+		return nil, &util.UnsupportedQueryError{
+			Message: "Plan is non-shared",
```

Comment:
i don't think that's a good error message, i issued a request for an instance and not for a plan. maybe something like "Cannot share instance of a non-shareable plan"?